homeassistant/auth/session.py

```diff
@@ -0,0 +1,199 @@
+        self._store = Store[StoreData](
+            hass, STORAGE_VERSION, STORAGE_KEY, private=True, atomic_writes=True
+        )
+        self._key: str | None = None
+        self._refresh_token_revoke_callbacks: dict[str, CALLBACK_TYPE] = {}
+
+    @property
+    def key(self) -> str:
+        """Return the encryption key."""
+        if self._key is None:
+            self._key = Fernet.generate_key().decode()
+            self._async_schedule_save()
+        return self._key
+
+    async def async_validate_strict_connection_session(
+        self,
+        request: Request,
+    ) -> bool:
+        """Check if a request has a valid strict connection session."""
+        session = await get_session(request)
+        if session.new:
+            return False
+        result = await self._async_validate_strict_connection_session(session)
+        if result is False:
+            session.invalidate()
+        return result
+
+    async def _async_validate_strict_connection_session(
+        self,
+        session: Session,
+    ) -> bool:
+        if not (session_id := session.get(SESSION_ID)):
+            return False
+
+        if token_id := self._strict_connection_sessions.get(session_id):
+            if self._auth.async_get_refresh_token(token_id):
+                return True
+            # refresh token is invalid, delete entry
+            self._strict_connection_sessions.pop(session_id)
+            self._async_schedule_save()
+
+        if data := self._temp_sessions.get(session_id):
+            if dt_util.utcnow() <= data.absolute_expiry:
+                return True
+            # session expired, delete entry
+            self._temp_sessions.pop(session_id).cancel_remove()
+
+        return False
+
+    @callback
+    def _async_register_revoke_token_callback(self, refresh_token_id: str) -> None:
+        """Register a callback to revoke all sessions for a refresh token."""
+        if refresh_token_id in self._refresh_token_revoke_callbacks:
+            return
+
+        @callback
+        def async_invalidate_auth_sessions() -> None:
+            """Invalidate all sessions for a refresh token."""
+            self._strict_connection_sessions = {
+                session_id: token_id
+                for session_id, token_id in self._strict_connection_sessions.items()
+                if token_id != refresh_token_id
+            }
+            self._async_schedule_save()
+
+        self._refresh_token_revoke_callbacks[refresh_token_id] = (
+            self._auth.async_register_revoke_token_callback(
+                refresh_token_id, async_invalidate_auth_sessions
+            )
+        )
+
+    async def async_create_session(
+        self,
+        request: Request,
+        refresh_token: RefreshToken,
+    ) -> None:
+        """Create new session for given refresh token."""
+        if not self._auth.async_get_refresh_token(refresh_token.id):
+            return
+
+        self._strict_connection_sessions = {
```

Comment:
I think we should add a comment here why we filter away all existing sessions for this refresh token. It means that there can only be one session per refresh token, right? This seems important.